### DISCUSSION TITLE
OneDrive integration: Trying to use LargeFileUploadTask with Proxy responds with 403

### DISCUSSION THREAD

**Discussion by GeertVerthe** [microsoftgraph/msgraph-sdk-dotnet#2603] [Q&A] [answered]
Hi, we are struggling to use the MS Graph nuget package in combination with a proxy. Our code runs perfectly with customers without a proxy configured. 

After struggling to get the authentication and normal MS Graph calls done succesfully through the nuget package, we are now facing our last burden: uploading the file through the proxy using LargeFileUploadTask.

We are using the OnBehalfOfCredential and are constructing our GraphServiceClient like this:

```
_proxy = new WebProxy("http://xxxxx:8080", true, null, CredentialCache.DefaultNetworkCredentials);
_httpClientHandler = new HttpClientHandler()
{
    UseProxy = true,
    UseDefaultCredentials = true,
    Proxy = _proxy 
};

var options = new OnBehalfOfCredentialOptions
{
    AuthorityHost = AzureAuthorityHosts.AzurePublicCloud,
    Transport = new HttpClientTransport(_httpClientHandler)
};
var onBehalfOfCredential = new OnBehalfOfCredential(ClientInfo.TenantID, ClientInfo.OboApp.ClientID, ClientInfo.OboApp.ClientSecret, ClientInfo.AadApp.AccessToken, options);

// Create an HttpClient using the GraphClientFactory
_httpClient = GraphClientFactory.Create(proxy: _belfiusproxy);
_graphServiceClient = new GraphServiceClient(_httpClient, onBehalfOfCredential, ClientInfo.OboApp.Scopes.Split(' '));
```

This does the trick for 'normal' calls for example, creating a OneDrive folder

Now for the LargeFileUploadTask, we can not make this work through the proxy. 

I understood, we should pass the IRequestAdaptor in the constructor of the LargeFileUploadTask but it does not work. 

Things we tried:
`var fileUploadTask = new LargeFileUploadTask<DriveItem>(uploadSession, fileStream, maxSliceSize, _graphServiceClient.RequestAdapter);`

and through looking at the internal code of the nuget package:

 ```
               var requestAdapter = new BaseGraphRequestAdapter(new AnonymousAuthenticationProvider(), httpClient: _httpClient)
                {
                    BaseUrl = uploadSession.UploadUrl
                };

                var fileUploadTask = new LargeFileUploadTask<DriveItem>(uploadSession, fileStream, maxSliceSize, requestAdapter);
```

but we keep on facing 403 - generalException errors.
![image](https://github.com/user-attachments/assets/56bfc091-3cf4-4314-9734-035aea3b8b51)

What do we do wrong?

Thanks in advance!

**Comment by andrueastman** [comment]
Thanks for raising this @GeertVerthe 

Are you by any chance able to make it work if you pass the `_graphServiceClient` to the task as below?

```cs
var fileUploadTask = new LargeFileUploadTask<DriveItem>(uploadSession, fileStream, maxSliceSize, _graphServiceClient);
```

**Reply by GeertVerthe** [reply]
Hi, thanks for your answer. I found indeed some topic in here somewhere that suggested that but it seems that the version of ms graph we are using (5.56.0 current latest) does not allow this...

![image](https://github.com/user-attachments/assets/62b88155-6c5a-4284-9ecc-2a0ba3df50ed)

I can use _graphServiceClient.RequestAdapter as the parameter, but as said in first thread, this does not apply the proxy it seems :(

**Reply by andrueastman** [reply]
Thanks for confirming.
Any chance you can confirm if there is a `RawResponseBody` property populated in the exception? Any chance you can share the contents?

**Reply by GeertVerthe** [reply, marked as answer]
Thanks for getting back to me. I currently have not access to the RawResponseBody but we are awaiting the customer to allow traffic to their https://[tenant].my-sharepoint.com as we can see this is the url that is used to perform the upload task. It might be that their proxy blocks this. 
Can you confirm that this url should be open? 

I tried with some sketchy free proxies found online, and these seem to work with my code, so I assume the problem lies with the customers proxy.

I'll get back to you whenever i get feedback from them!

**Reply by GeertVerthe** [reply]
Sorry for the delay but whitelisting the xxx.my-sharepoint.com url on their proxy fixed the issue!
